celery/canvas.py

```diff
@@ -395,23 +395,41 @@ def __or__(self, other):
             other = maybe_unroll_group(other)
             if isinstance(self, _chain):
                 # chain | group() -> chain
-                sig = self.clone()
-                sig.tasks.append(other)
-                return sig
+                return _chain(
+                    seq_concat_item([
+                        reduce(
```

Comment:
Can you please explain the logic here? I cannot figure out the use of `reduce`. It seems that the concept is to have the next task signature become the error callback?

Reply:
With `reduce` I'm assigning tasks from `link_error` of current chain to individual tasks with respect to their order.
This code equals to:

    new_tasks = []
    for task in self.tasks:
      new_task = task.clone()
      new_tasks.append(new_task)
      for error_handler in self._with_list_option('link_error'):
        new_task.on_error(error_handler)
    return _chain(seq_concat_item(new_tasks, other), app=self._app)`

Reply:
I would go with this version since it is far more readable. Also, I think that using `_with_list_option` is not required.

```
tasks = [task.clone() for task in self.tasks]
error_handler = self.options.get('link_error')
if error_handler is not None:   
    for task in tasks:
        task.on_error(error_handler)
```

What do you think?

Reply:
OK, I'll update my branch.
I've used `_with_list_option` because `link_error` is a list of tasks and there are some logic on setting it in case of `MutableSequence`.
But here we just reading it, so I can use `.options.get('link_error', [])`